File: qlat/pylib/qlat/scalar_action.py
import qlat.c as c

from qlat.field import *
from qlat.field_utils import *

class ScalarAction:

    def __init__(self, m_sq, lmbd, alpha):
        self.cdata = c.mk_scalar_action(m_sq, lmbd, alpha)

    def __del__(self):
        assert isinstance(self.cdata, int)
        c.free_scalar_action(self)

    def __imatmul__(self, v1):
        assert isinstance(v1, ScalarAction)
        c.set_scalar_action(self, v1)
        return self

    def m_sq(self):
        return c.get_m_sq_scalar_action(self)

    def lmbd(self):
        return c.get_lmbd_scalar_action(self)

    def lmbd(self):
        return c.get_alpha_scalar_action(self)

    def action_node(self, sf):
        assert isinstance(sf, FieldBase)
        return c.action_node_scalar_action(self, sf)

    def hmc_estimate_mass(self, masses, field_ft, force_ft, phi0):
        assert isinstance(masses, FieldBase)
        assert isinstance(field_ft, FieldBase)
        assert isinstance(force_ft, FieldBase)
        return c.hmc_estimate_mass_scalar_action(self, masses, field_ft, force_ft, phi0)

    def to_mass_factor(self, sin_domega):
        assert isinstance(sin_domega, FieldBase)
        return c.to_mass_factor_scalar_action(self, sin_domega)

    def set_complex_from_double(self, cf, sf):
        assert isinstance(cf, FieldBase)
        assert isinstance(sf, FieldBase)
        return c.set_complex_from_double_scalar_action(self, cf, sf)

    def set_double_from_complex(self, sf, cf):
        assert isinstance(cf, FieldBase)
        assert isinstance(sf, FieldBase)
        return c.set_double_from_complex_scalar_action(self, sf, cf)

    def sum_sq(self, sf):
        assert isinstance(sf, FieldBase)
        return c.sum_sq_scalar_action(self, sf)

    def hmc_m_hamilton_node(self, sf, masses):
        assert isinstance(sf, FieldBase)
        assert isinstance(masses, FieldBase)
        return c.hmc_m_hamilton_node_scalar_action(self, sf, masses)

    def hmc_set_force(self, sm_force, sf):
        assert isinstance(sm_force, FieldBase)
        assert isinstance(sf, FieldBase)
        return c.hmc_set_force_scalar_action(self, sm_force, sf)

    def hmc_field_evolve(self, sf_ft, sm_ft, masses, step_size):
        assert isinstance(sf_ft, FieldBase)
        assert isinstance(sm_ft, FieldBase)
        assert isinstance(masses, FieldBase)
        c.hmc_field_evolve_scalar_action(self, sf_ft, sm_ft, masses, step_size)

    def axial_current_node(self, axial_current, sf):
        assert isinstance(sf, FieldBase)
        assert isinstance(axial_current, FieldBase)
        return c.axial_current_node_scalar_action(self, axial_current, sf)

    def hmc_set_rand_momentum(self, sm_complex, masses, rs):
        assert isinstance(sm_complex, FieldBase)
        assert isinstance(masses, FieldBase)
        assert isinstance(rs, RngState)
        return c.hmc_set_rand_momentum_scalar_action(self, sm_complex, masses, rs)

    def hmc_predict_field(self, field_ft, momentum_ft, masses, vev_sigma):
        assert isinstance(field_ft, FieldBase)
        assert isinstance(momentum_ft, FieldBase)
        assert isinstance(masses, FieldBase)
        return c.hmc_predict_field_scalar_action(self, field_ft, momentum_ft, masses, vev_sigma)

    def get_polar_field(self, polar_field, field):
        assert isinstance(polar_field, FieldBase)
        assert isinstance(field, FieldBase)
        return c.get_polar_field_scalar_action(self, polar_field, field)

File: qlat/pylib/cqlat/scalar-action.cpp
#include "lib.h"

EXPORT(mk_scalar_action, {
  using namespace qlat;
  double m_sq = 1.0;
  double lmbd = 1.0;
  double alpha = 0.0;
  if (!PyArg_ParseTuple(args, "d|d|d", &m_sq, &lmbd, &alpha)) {
    return NULL;
  }
  ScalarAction* psa = new ScalarAction(m_sq, lmbd, alpha);
  return py_convert((void*)psa);
})

EXPORT(free_scalar_action, {
  using namespace qlat;
  return free_obj<ScalarAction>(args);
})

EXPORT(set_scalar_action, {
  using namespace qlat;
  return set_obj<ScalarAction>(args);
})

EXPORT(get_m_sq_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_sa)) {
    return NULL;
  }
  const ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  return py_convert(sa.m_sq);
})

EXPORT(get_lmbd_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_sa)) {
    return NULL;
  }
  const ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  return py_convert(sa.lmbd);
})

EXPORT(get_alpha_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_sa)) {
    return NULL;
  }
  const ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  return py_convert(sa.alpha);
})

EXPORT(action_node_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_sf = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_sa, &p_sf)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  const Field<double>& sf = py_convert_type<Field<double>>(p_sf);
  const double ret = sa.action_node(sf);
  return py_convert(ret);
})

EXPORT(hmc_estimate_mass_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_masses = NULL;
  PyObject* p_field_ft = NULL;
  PyObject* p_force_ft = NULL;
  double phi0 = 0.0;
  if (!PyArg_ParseTuple(args, "OOOOd", &p_sa, &p_masses, &p_field_ft, &p_force_ft, &phi0)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<double>& masses = py_convert_type<Field<double>>(p_masses);
  const Field<Complex>& field_ft = py_convert_type<Field<Complex>>(p_field_ft);
  const Field<Complex>& force_ft = py_convert_type<Field<Complex>>(p_force_ft);
  sa.hmc_estimate_mass(masses, field_ft, force_ft, phi0);
  Py_RETURN_NONE;
})

EXPORT(to_mass_factor_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_sin_domega = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_sa, &p_sin_domega)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<double>& sin_domega = py_convert_type<Field<double>>(p_sin_domega);
  sa.to_mass_factor(sin_domega);
  Py_RETURN_NONE;
})

EXPORT(hmc_m_hamilton_node_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_sm = NULL;
  PyObject* p_masses = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_sa, &p_sm, &p_masses)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  const Field<Complex>& sm = py_convert_type<Field<Complex>>(p_sm);
  const Field<double>& masses = py_convert_type<Field<double>>(p_masses);
  const double ret = sa.hmc_m_hamilton_node(sm, masses);
  return py_convert(ret);
})

EXPORT(hmc_set_force_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_sm_force = NULL;
  PyObject* p_sf = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_sa, &p_sm_force, &p_sf)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<double>& sm_force = py_convert_type<Field<double>>(p_sm_force);
  const Field<double>& sf = py_convert_type<Field<double>>(p_sf);
  sa.hmc_set_force(sm_force, sf);
  Py_RETURN_NONE;
})

EXPORT(hmc_field_evolve_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_sf = NULL;
  PyObject* p_sm = NULL;
  PyObject* p_masses = NULL;
  double step_size = 0.0;
  if (!PyArg_ParseTuple(args, "OOOOd", &p_sa, &p_sf, &p_sm, &p_masses, &step_size)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<Complex>& sf = py_convert_type<Field<Complex>>(p_sf);
  const Field<Complex>& sm = py_convert_type<Field<Complex>>(p_sm);
  const Field<double>& masses = py_convert_type<Field<double>>(p_masses);
  sa.hmc_field_evolve(sf, sm, masses, step_size);
  Py_RETURN_NONE;
})

EXPORT(axial_current_node_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_axial_cur = NULL;
  PyObject* p_sf = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_sa, &p_axial_cur, &p_sf)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<double>& axial_cur = py_convert_type<Field<double>>(p_axial_cur);
  const Field<double>& sf = py_convert_type<Field<double>>(p_sf);
  sa.axial_current_node(axial_cur, sf);
  Py_RETURN_NONE;
})

EXPORT(sum_sq_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_f = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_sa, &p_f)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  const Field<double>& f = py_convert_type<Field<double>>(p_f);
  const double ret = sa.sum_sq(f);
  return py_convert(ret);
})

EXPORT(hmc_set_rand_momentum_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_sm = NULL;
  PyObject* p_masses = NULL;
  PyObject* p_rs = NULL;
  if (!PyArg_ParseTuple(args, "OOOO", &p_sa, &p_sm, &p_masses, &p_rs)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<Complex>& sm = py_convert_type<Field<Complex>>(p_sm);
  Field<double>& masses = py_convert_type<Field<double>>(p_masses);
  const RngState& rs = py_convert_type<RngState>(p_rs);
  sa.hmc_set_rand_momentum(sm, masses, rs);
  Py_RETURN_NONE;
})

EXPORT(hmc_predict_field_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_sf_ft = NULL;
  PyObject* p_sm_ft = NULL;
  PyObject* p_masses = NULL;
  double vev_sigma = 0.0;
  if (!PyArg_ParseTuple(args, "OOOOd", &p_sa, &p_sf_ft, &p_sm_ft, &p_masses, &vev_sigma)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<Complex>& sf_ft = py_convert_type<Field<Complex>>(p_sf_ft);
  const Field<Complex>& sm_ft = py_convert_type<Field<Complex>>(p_sm_ft);
  const Field<double>& masses = py_convert_type<Field<double>>(p_masses);
  sa.hmc_predict_field(sf_ft, sm_ft, masses, vev_sigma);
  Py_RETURN_NONE;
})

EXPORT(get_polar_field_scalar_action, {
  using namespace qlat;
  PyObject* p_sa = NULL;
  PyObject* p_pf = NULL;
  PyObject* p_sf = NULL;
  if (!PyArg_ParseTuple(args, "OOO", &p_sa, &p_pf, &p_sf)) {
    return NULL;
  }
  ScalarAction& sa = py_convert_type<ScalarAction>(p_sa);
  Field<double>& pf = py_convert_type<Field<double>>(p_pf);
  const Field<double>& sf = py_convert_type<Field<double>>(p_sf);
  sa.get_polar_field(pf, sf);
  Py_RETURN_NONE;
})
